cluster-operator/src/main/java/io/strimzi/operator/cluster/model/AbstractModel.java

```diff
@@ -973,11 +973,16 @@ protected void heapOptions(List<EnvVar> envVars, double dynamicHeapFraction, lon
             // Honour explicit max heap
             kafkaHeapOpts.append(' ').append("-Xmx").append(xmx);
         } else {
-            // Otherwise delegate to the container to figure out
-            // Using whatever cgroup memory limit has been set by the k8s infra
-            envVars.add(buildEnvVar(ENV_VAR_DYNAMIC_HEAP_FRACTION, Double.toString(dynamicHeapFraction)));
-            if (dynamicHeapMaxBytes > 0) {
-                envVars.add(buildEnvVar(ENV_VAR_DYNAMIC_HEAP_MAX, Long.toString(dynamicHeapMaxBytes)));
+            Resources resources = getResources();
+            CpuMemory cpuMemory = resources == null ? null : resources.getLimits();
+            // Delegate to the container to figure out only when cgroup memory limits are defined to prevent allocating
+            // too much memory on the kubelet.  When no memory limit is defined let JVM allocate "just enough" memory
+            // by not trying to guess `Xms` or `Xmx`.
+            if (resources != null && cpuMemory != null && cpuMemory.getMemory() != null) {
```

Comment:
When I looked at write usages of the `resources` member I only found a setter, but no guarantee that it is actually defined so I thought it would be prudent to be defensive.  WDYT?

Reply:
I'm not quite sure what you mean. I mean that you've defined `CpuMemory cpuMemory = resources == null ? null : resources.getLimits();` in the statement above, so it's sufficient that `cpuMemory != null` because if it is nonnull then resources must also be nonnull.

Reply:
I see what you mean now.  You're right.  I'll update.